### ISSUE TITLE
RSA "raw" decryption does not honor block size

### ISSUE THREAD

**Issue by lapo-luchini** [bcgit/bc-java#1017]
It seems that the following code fails about 1 time in 256.

```Java
Cipher c = Cipher.getInstance("RSA/ECB/NoPadding", bc);
c.init(Cipher.DECRYPT_MODE, key);
byte[] raw = c.doFinal(padded);
if (raw.length != padded.length) throw new Exception();
```

Full test-case here: [TestRSA0x00_RawDecrypt.zip](https://github.com/bcgit/bc-java/files/7100296/TestRSA0x00_RawDecrypt.zip)

RSA "raw" is defined on numbers, not on byte arrays, so this is maybe not strictly wrong to do but I see no reason to return an array smaller than a logical block [iff the first byte is zero](https://github.com/bcgit/bc-java/blob/930f8b274c4f1f3a46e68b5441f1e7fadb57e8c1/core/src/main/java/org/bouncycastle/crypto/engines/RSACoreEngine.java#L152) (strangely enough, any second or third zero byte is kept as is), furthermore at the cost of artificially slowing down even [the common case](https://github.com/bcgit/bc-java/blob/930f8b274c4f1f3a46e68b5441f1e7fadb57e8c1/core/src/main/java/org/bouncycastle/crypto/engines/RSACoreEngine.java#L159) to maintain equal speed in all branches.

It could probably make sense to drop all that code and simply return the full block, which is the same behavior of SunJCE provider.

PS: Bouncy Castle also recognizes the name `"RSA/None/NoPadding"` (and has the same behavior) for which a shorter output would seem a bit more normal to me, but for the name `"RSA/ECB/NoPadding"` (also recognized by SunJCE) IMvHO honoring the natural block size is the behavior assumed by the user (or, at the very least, it is the behavior assumed by me 😇), given the 'B' in ECB.

**Comment by dghgit** [comment]
This is a case of garbage in garbage out. Assuming octet boundaries, the logical block size of an RSA cipher, in as much as one could exist, is (modulus.bitLength() + 7) / 8 - 1 as anything bigger might end up being outside of the group defined by the keys.

**Comment by alexdupre** [comment]
I'm not sure it's correct to assume octet boundaries. The cryptographic primitive accepts `an integer between 0 and n - 1`, so the RSA core implemenation shouldn't exclude an entire byte, but simply checks that it's within the modulus.
And the decryption, given that returns a `byte[]` and not a `BigInteger` shouldn't trim the first null byte (there may be zero, one or more).

**Comment by mouse07410** [comment]
I concur - it does seem wrong to subtract 1 in that formula.

Plus, the whole computation is weird:
- if octet boundary is enforced `modulus.bitLength()/8`;
- otherwise `(modulus.bitLength()+7)/8`.

**Comment by dghgit** [comment]
I'm beginning to think this whole pandemic thing is starting to get the better of us, myself included.

Let me try again. In some ways this is an age old debate, but if I make the change requested, the following will no longer work:

[RSARawSample.zip](https://github.com/bcgit/bc-java/files/7105885/RSARawSample.zip)

We made the original call the way we did as we took unpadded as "nothing added". 

The reason for padding, of course with RSA, is to ensure "nothing dropped", but at any rate for some 21 years people have been happily encrypting and decrypting data that doesn't have leading zeroes, if I make the change requested that will no longer work. I don't believe there is a right or a wrong position with this one, but it should be obvious why I would not want to make the change suggested.

**Comment by alexdupre** [comment]
I agree that there isn't a right or wrong position, and it's understandable that a change might cause more breakages than fixes, I think here we wanted to understand the reason of such choice.
Your example works fine with the SunJCE provider, though, that doesn't strip the leading zero. And the reason seems to be that the "RSA" algorithm is mapped to RSA/PKCS1 padding in the Sun provider, while it's mapped to RSA/NoPadding in BC. Changing it would definitively break things.

**Comment by dghgit** [comment]
Yes, if you specify "RSA", "AES", or anything else without qualification, what you get back is platform dependent. As you've noted the point really was more about raw RSA behavior with BC.
